src/plugins/s390.c

```diff
@@ -398,3 +399,353 @@ gchar* bd_s390_zfcp_sanitize_lun_input (const gchar *lun, GError **error) {
 
     return fulllun;
 }
+
+/**
+ * bd_s390_zfcp_online:
+ * @devno: zfcp device number
+ * @wwpn: zfcp WWPN (World Wide Port Number)
+ * @lun: zfcp LUN (Logical Unit Number)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether a zfcp device was successfully switched online
+ */
+gboolean bd_s390_zfcp_online (const gchar *devno, const gchar *wwpn, const gchar *lun, GError **error) {
+    gboolean boolrc = FALSE;
+    gint rc = 0;
+    FILE *fd = NULL;
+    DIR *pdfd = NULL;
+    const gchar *zfcp_cio_free[4] = {"/usr/sbin/zfcp_cio_free", "-d", devno, NULL};
+    const gchar *chccwdev[4] = {"/usr/sbin/chccwdev", "-e", devno, NULL};
+
+    gchar *zfcpsysfs = "/sys/bus/ccw/drivers/zfcp";
+    gchar *online = g_strdup_printf ("%s/%s/online", zfcpsysfs, devno);
+    gchar *portdir = g_strdup_printf ("%s/%s/%s", zfcpsysfs, devno, wwpn);
+    gchar *unitadd = g_strdup_printf ("%s/unit_add", portdir);
+    gchar *unitdir = g_strdup_printf ("%s/%s", portdir, lun);
+    gchar *failed = g_strdup_printf ("%s/failed", unitdir);
+
+    /* part 01: make sure device is available/not on device ignore list */
+    fd = fopen (online, "r");
+    if (!fd) {
+        boolrc = bd_utils_exec_and_report_error (zfcp_cio_free, NULL, error);
+
+        if (!boolrc) {
+            fclose (fd);
+            g_free (online);
```

Comment:
This should free a lot more strings. Maybe allocate them right before they are actually needed?